Adds PR template

KipHamiltons opened the pull request:
### Pull Request Template 

This adds a template for making a PR.

### Reviewers, Note

This PR was made using the template

### Things left to do

- [x] Nothing

ysims reviewed (commented):
Add in checks for 'I have updated any relevent module README', 'I have updated NUbook if necessary', and 'I have added/updated tests if necessary'?

CMurtagh-LGTM commented:
This shouldn't be a tutorial, rather it should be things that should go in the actual PR or things people may forget.

LachlanCourt commented:
Maybe add something about how to test, perhaps in the Reviewers, Note section?

ysims commented:
> This shouldn't be a tutorial, rather it should be things that should go in the actual PR or things people may forget.

He has a point. Maybe just link to a NUbook page?

CMurtagh-LGTM commented:
> Maybe add something about how to test, perhaps in the Reviewers, Note section?

Its good to have this somewhere, but I feel this is not the right place.

ysims commented:
> > Maybe add something about how to test, perhaps in the Reviewers, Note section?
> 
> Its good to have this somewhere, but I feel this is not the right place.

I agree with @LachlanCourt - I think you might misunderstand what he means @CMurtagh-LGTM. 
Specifying how reviewers can test this particular PR is good.

KipHamiltons commented:
I removed the reviewers part because it was more of a tutorial (as per @CMurtagh-LGTM's suggestion), and added a checklist with things to make sure you do (as suggested by @ysims). I also specified where to tell reviewers how to verify/test the changes (thanks @LachlanCourt).
Thanks for the suggestions guys. Great points. What do you think of the changes?

LachlanCourt commented on .github/pull_request_template.md:
```diff
@@ -5,11 +5,19 @@ Things the PR does go here
+- [ ] Added/Updated tests for your changes, including regression tests for bug fixes
+- [ ] Updated relevant module READMEs
+
 ## Optional Headings
```
Should this be a heading 3?

KipHamiltons replied:
It's a higher level than the others that follow it, because they're all _optional_, so I deliberately made it a bigger heading.

LachlanCourt replied:
Gotcha sorry I thought that was an example of how to add a custom optional heading

ysims replied:
Having a level 2 heading might be an accessibility issue. I feel like GitHub should handle this itself, but PR comments live under a level 2 heading. It might be confusing to have this level 2 here.

KipHamiltons replied:
I want to communicate to people that they should remove the headings they don't use, and they should also remove that heading `## Optional Headings` in all cases. I am not sure how to go about that though. Do you have any suggestions?

ysims replied:
I'd suggest increasing the heading level of all of them by one. Maybe add in a line under this heading saying something along the lines of 'delete anything in this section that is not needed (including this line)'.

ysims replied:
Oh wait but they're removing the heading. Maybe use a line break.

LachlanCourt reviewed (approved):
I'm happy if you fix dependabot :D

KipHamiltons commented:
You're fixing dependabot in your branch though @LachlanCourt  :p

LachlanCourt commented:
Approve my branch then :p
But yea off you go XD

Bidski commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,40 @@
+### Pre-PR Checklist:
+
+Have you
+- [ ] Updated NUbook if necessary
```
```suggestion
- [ ] Updated NUbook if necessary (add link to NUbook PR here or commit hash)
```

Bidski reviewed (changes requested):
Should we also have an issues template?

KipHamiltons commented:
> Should we also have an issues template?

Yes. We should.

ysims commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,40 @@
+
+**If there are many things left to do then consider a [_draft_ PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/)**
+
+## Labels Info (delete the rest of this once you have read it)
```
Maybe this stuff can just go into the Pre-PR Checklist in one line, like 
- [ ] [Added relevant labels to the PR](https://github.com/NUbots/NUbots/labels)

KipHamiltons replied:
I like the idea of adding this as a todo item as you say, so I have added that. However, I think it's worth documenting the label system here, because this is the place you're going to be wondering most about how the labels are organised.

ysims replied:
I think it'd be good to add in a link to all the labels (https://github.com/NUbots/NUbots/labels)

CMurtagh-LGTM commented:
@KipHamiltons Check the history, I think you may of made a mistake

KipHamiltons commented:
Github did not like my (accidental) local rebase and push lol. yikes. Thanks for pointing that out @CMurtagh-LGTM

BryceTuppurainen commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,41 @@
+- [ ] Added/Updated tests for your changes, including regression tests for bug fixes
+- [ ] Updated relevant module READMEs
+- [ ] Added relevant labels to the PR
+
```
```suggestion
- [ ] Ensured that the PR builds successfully (or issues are identified here)
- [ ] [Formatted any documents edited](https://nubook.nubots.net/guides/general/git#formatting-and-buildkite-checks)
- [ ] [Added any required directives in documentation](https://nubook.nubots.net/guides/general/documentation#extra-formatting)

```

I agree with what Cameron suggested earlier this doesn't need to be a tutorial for how to create a PR and acts only to confirms any necessary info exists and cleans it up consistently.

Just figured particularly with newer members this info and links in the checklist may be useful (also if we mix up our documentation style in future or otherwise change formatting etc linking over to the NUbook page will help keep this up to date and make it easy to jump over to in order to check the current standards)

KipHamiltons replied:
I personally don't think we should add things which are caught by CI like formatting and builds to the template - I reckon we should avoid making this template too large, lest people not bother reading it (it's probably larger than I would like already :( )
I reckon that the directives are probably worth adding though :+1: good idea.

Regarding difficulty labels, I had originally intended them to be for issues and things which you were looking for other people to contribute on. In this example, although this PR isn't "hard" to contribute to, it's not something I especially want someone to jump in and contribute to because they saw it's "easy" for them to help with, which was the original intention of those labels. If you think those labels could be improved with a different conceptualisation though, I am all ears.

ysims commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,44 @@
+
+**If there are many things left to do then consider a [_draft_ PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/)**
+
+## Labels Info (delete the rest of this once you have read it)
```
This could just be
```suggestion
### Labels Info (delete the rest of this once you have read it)
```
now, right?

ysims reviewed (approved):
Yeah. Can see how it goes with actually using it as well and can adjust if needed.
